File: 04_Binary_Search/1D Arrays/Python/01_binary_search.py
def binary_search(arr: list, target: int):
  low = 0
  high = len(arr) - 1

  while low <= high:
    mid = int( low + (high-low)/2 )

    if arr[mid] == target:
      return mid
    elif arr[mid] > target:
      high = mid - 1
    else:
      low = mid + 1
  return -1


def bs_algo(arr: list, low: int, high: int, target: int):
  # base condition
  if low > high:
    return -1
  

  mid = int(low + (high-low)/2)

  if arr[mid] == target:
    return mid
  elif arr[mid] > target:
    return bs_algo(arr, low, mid - 1, target)
  else:
    return bs_algo(arr, mid + 1, high, target)


def binary_search_recursive(arr: list, target: int):
  low = 0
  high = len(arr) - 1
  return bs_algo(arr, low, high, target)

size = int(input("Enter size: "))
arr = list(map(int, input("Enter elements in array in sorted manner: ").strip().split()[:size]))

target = int(input("Enter the target: "))

# target_index = binary_search(arr, target)
target_index = binary_search_recursive(arr, target)

if target_index >= 0:
  print("Target found at index: ", target_index)
else:
  print("Target NOT found")

File: 04_Binary_Search/1D Arrays/C++/01_binary_search.cpp
// given -> 1) sorted array
//          2) target to find in the array

#include <bits/stdc++.h>
using namespace std;

// it assumess the array is already sorted
int binarySearch(vector<int> &arr, int target){
  int low = 0;
  int high = arr.size() - 1;

  // it uses bisection method
  while( low <= high ){
    int mid = low + (high - low)/2;
    
    if( arr[mid] == target )
      return mid;

      // if mid element is > target
      //    -> target might be before the mid
      //    -> so update high
    else if( arr[mid] > target )
            high = mid - 1;

      // else if mid element is < target
      //    -> target might be after the mid
      //    -> update the loq
    else low = mid + 1;
  }

  // if target not found
  return -1;
}

int bs_algo(vector<int>&arr, int low, int high, int target){
  // base condition
  if( low > high ) return -1;


  int mid = low + (high-low)/2;

  if(arr[mid] == target)
    return mid;
  else if(arr[mid] > target)
    return bs_algo(arr, low, mid - 1, target);
  else return bs_algo(arr, mid + 1, high, target);

}

int binarySearch_recursive(vector<int>& arr, int target){
  int low = 0;
  int high = arr.size() - 1;

  return bs_algo(arr, low, high, target);
}

int main(int argc, char* argv[]){
  vector<int> v;

  int size;
  cin >> size;

  // enter sorted array
  for(int i=0; i<size; ++i){
    int a;
    cin >> a;
    v.push_back(a);
  }

  int target;
  cin >> target;

  // int target_index = binarySearch(v, target);
  int target_index = binarySearch_recursive(v, target);

  if(target_index >= 0)
    cout << "Target found at index: " << target_index;
  else
    cout << "Target NOT found";
return 0;
}